crates/sparrow-instructions/src/evaluators/list/collect_primitive.rs

```diff
@@ -0,0 +1,131 @@
+        };
+
+        let max = match info.args[0].value_ref.literal_value() {
+            Some(ScalarValue::Int64(Some(v))) if *v <= 0 => {
+                anyhow::bail!("unexpected value of `max` -- must be > 0")
+            }
+            Some(ScalarValue::Int64(Some(v))) => *v as usize,
+            // If a user specifies `max = null`, we use usize::MAX value as a way
+            // to have an "unlimited" buffer.
+            Some(ScalarValue::Int64(None)) => usize::MAX,
+            Some(other) => anyhow::bail!("expected i64 for max parameter, saw {:?}", other),
+            None => anyhow::bail!("expected literal value for max parameter"),
+        };
+
+        let (_, input, tick, duration) = info.unpack_arguments()?;
+        Ok(Box::new(Self {
+            max,
+            input,
+            tick,
+            duration,
+            token: CollectToken::default(),
+        }))
+    }
+}
+
+impl<T> Evaluator for CollectPrimitiveEvaluator<T>
+where
+    T: ArrowPrimitiveType + Send + Sync,
+    T::Native: Serialize + DeserializeOwned + Copy,
+{
+    fn evaluate(&mut self, info: &dyn RuntimeInfo) -> anyhow::Result<ArrayRef> {
+        match (self.tick.is_literal_null(), self.duration.is_literal_null()) {
+            (true, true) => self.evaluate_non_windowed(info),
+            (true, false) => unimplemented!("since window aggregation unsupported"),
+            (false, false) => panic!("sliding window aggregation should use other evaluator"),
+            (_, _) => anyhow::bail!("saw invalid combination of tick and duration"),
+        }
+    }
+
+    fn state_token(&self) -> Option<&dyn StateToken> {
+        Some(&self.token)
+    }
+
+    fn state_token_mut(&mut self) -> Option<&mut dyn StateToken> {
+        Some(&mut self.token)
+    }
+}
+
+impl<T> CollectPrimitiveEvaluator<T>
+where
+    T: ArrowPrimitiveType + Send + Sync,
+    T::Native: Serialize + DeserializeOwned + Copy,
+{
+    fn ensure_entity_capacity(&mut self, len: usize) {
+        self.token.resize(len)
+    }
+
+    fn evaluate_non_windowed(&mut self, info: &dyn RuntimeInfo) -> anyhow::Result<ArrayRef> {
+        let input = info.value(&self.input)?.array_ref()?;
+        let key_capacity = info.grouping().num_groups();
+        let entity_indices = info.grouping().group_indices();
+        assert_eq!(entity_indices.len(), input.len());
+
+        self.ensure_entity_capacity(key_capacity);
+
+        let input = downcast_primitive_array::<T>(input.as_ref())?;
+        let builder = PrimitiveBuilder::<T>::new();
+        let mut list_builder = ListBuilder::new(builder);
+
+        izip!(entity_indices.values(), input).for_each(|(entity_index, input)| {
+            let entity_index = *entity_index as usize;
+
+            self.token.add_value(self.max, entity_index, input);
+            let cur_list = self.token.state(entity_index);
+
+            list_builder.append_value(cur_list.clone());
+        });
```

Comment:
I expect there's a cleverer way to do this by creating the list directly using the offsets and `take`